File: week1/GamGomYang/BOJ10845.c
#include <stdio.h>
#include <string.h>

#define MAXNUMBER 10000

typedef struct {
    int data[MAXNUMBER];
    int front;
    int rear;
} Queue;

void init(Queue *q) {
    q->front = 0;
    q->rear = 0;
}

int empty(Queue *q) {
    return q->rear == q->front;
}

void push(Queue *q, int value) {
    q->data[q->rear] = value;
    q->rear++;
}

int pop(Queue *q) {
    if (empty(q)) {
        return -1;
    } else {
        return q->data[q->front++];
    }
}

int front(Queue *q) {
    if (empty(q)) {
        return -1;
    } else {
        return q->data[q->front];
    }
}

int back(Queue *q) {
    if (empty(q)) {
        return -1;
    } else {
        return q->data[q->rear - 1];
    }
}

int size(Queue *q) {
    return q->rear - q->front;
}

int isEmpty(Queue *q) {
    return empty(q);
}

int main() {
    int n;
    char command[6];
    int value;

    Queue q;

    init(&q);

    scanf("%d", &n);

    for (int i = 0; i < n; i++) {
        scanf("%s", command);
        if (strcmp(command, "push") == 0) {
            scanf("%d", &value);
            push(&q, value);
        } else if (strcmp(command, "pop") == 0) {
            printf("%d\n", pop(&q));
        } else if (strcmp(command, "size") == 0) {
            printf("%d\n", size(&q));
        } else if (strcmp(command, "empty") == 0) {
            printf("%d\n", isEmpty(&q));
        } else if (strcmp(command, "front") == 0) {
            printf("%d\n", front(&q));
        } else if (strcmp(command, "back") == 0) {
            printf("%d\n", back(&q));
        }
    }

    return 0;
}

File: week1/GamGomYang/BOJ10845.py
from collections import deque
import sys

input = sys.stdin.read
data = input().split()

n = int(data[0])
queue = deque()
result = []

index = 1

for _ in range(n):
    command = data[index]
    if command == "push":
        index += 1
        value = int(data[index])
        queue.append(value)

    elif command == "pop":
        if queue:
            result.append(queue.popleft())
        else:
            result.append(-1)


    elif command == "size":
        result.append(len(queue))

    elif command == "empty":
        result.append(1 if not queue else 0)

    elif command == "front":
        if queue:
            result.append(queue[0])
        else:
            result.append(-1)


    elif command == "back":
        if queue:
            result.append(queue[-1])
        else:
            result.append(-1)
            
    
    index += 1

print("\n".join(map(str, result)))
